File: cpp/src/Slice.cpp
/*
 * Slice.cpp
 *
 *  Created on: Oct 21, 2015
 *      Author: D.J
 */

#include <vector>
#include <string>
#include <iostream>
#include <sstream>

#include "Slice.hpp"

using namespace std;

Slice::Slice(string values, int width) {
    m_width = width;

    // Storage for row
    CubieVector row;

    for (size_t i = 0; i < values.length(); ++i) {
        row.push_back(values.substr(i, 1));

        // End of row
        if ((i + 1) % width == 0) {
            m_cubies.push_back(row);
            row.clear();
        }
    }
}

Slice::Slice(CubieMatrix cubies) {
    m_width = cubies.size();
    m_cubies = cubies;
}

Slice::~Slice() { }

CubieVector Slice::row(int row) {
    return m_cubies[row];
}

bool Slice::rotate(Rotation rot) {
    switch (rot) {
        case CLOCKWISE:
            m_cubies = _reverse(_transpose(m_cubies), HORIZONTAL);

            break;
        case COUNTER_CLOCKWISE:
            m_cubies = _reverse(_transpose(m_cubies), VERTICAL);

            break;
        case HALF:
            m_cubies = _reverse(_reverse(m_cubies, HORIZONTAL), VERTICAL);

            break;
    }

    return true;
}

string Slice::format() {
    stringstream format;

    for (size_t i = 0; i < m_cubies.size(); ++i)
        format << formatRow(i) << "\n";

    return format.str();
}

CubieMatrix Slice::_transpose(CubieMatrix cubies) {
    CubieMatrix transposedCubies = cubies;

    for (size_t i = 0; i < cubies.size(); ++i)
        for (size_t j = 0; j < cubies[i].size(); ++j)
            transposedCubies[j][i] = cubies[i][j];

    return transposedCubies;
}

string Slice::formatRow(int row) {
    stringstream format;

    for (size_t i = 0; i < m_cubies[row].size(); ++i)
        format << " " << m_cubies[row][i].value();

    return format.str();
}

CubieMatrix Slice::_reverse(CubieMatrix cubies, Axis axis) {
    int width = cubies.size();
    CubieMatrix reversedCubies = cubies;

    switch (axis) {
        case HORIZONTAL:
            for (int i = 0; i < width; ++i)
                for (int j = 0; j < width; ++j)
                    reversedCubies[i][width - j - 1] = cubies[i][j];

            break;
        case VERTICAL:
            for (int i = 0; i < width; ++i)
                for (int j = 0; j < width; ++j)
                    reversedCubies[width - i - 1][j] = cubies[i][j];

            break;
    }

    return reversedCubies;
}

CubieVector Slice::column(int col) {
    return _transpose(m_cubies)[col];
}

int Slice::width() const {
    return m_width;
}

bool Slice::setRow(CubieVector row, int idx) {
    m_cubies[idx] = row;

    return true;
}

bool Slice::setColumn(CubieVector column, int idx) {
    m_cubies = _transpose(m_cubies);
    m_cubies[idx] = column;
    m_cubies = _transpose(m_cubies);

    return true;
}

std::string Slice::state() {
    stringstream format;

    for (size_t i = 0; i < m_cubies.size(); ++i)
        for (size_t j = 0; j < m_cubies[i].size(); ++j)
            format << m_cubies[i][j].value();

    return format.str();
}

File: php/src/Components/Slice.php
<?php
namespace Zinc\Components;

abstract class Rotation extends \Zinc\Lib\BasicEnum {
    const CLOCKWISE         = 0;
    const COUNTER_CLOCKWISE = 1;
    const HALF              = 2;
}

abstract class Axis extends \Zinc\Lib\BasicEnum {
    const HORIZONTAL = 0;
    const VERTICAL   = 1;
}

class Slice {
    // Storage for slice width
    // (all slices in cube should be same width)
    private $m_width;
    
    // Storage for cubies in slice
    private $m_cubies;
    
    public function __construct($values, $width = 0) {
        $this->m_cubies = new CubieMatrix;
        
        if ($values instanceof CubieMatrix) {
            $this->m_width = $values->count();
            $this->m_cubies = $values;
        }
        
        else {
            $this->m_width = $width;
            
            // Split string into rows
            $rows = str_split($values, $width);
            
            // Iterate rows
            foreach ($rows as $rowString) {
                // Add row vector to matrix
                $this->m_cubies[] = new CubieVector(str_split($rowString));
            }
        }
    }
    
    // Return slice width
    public function width() {
        return $this->m_width;
    }
    
    // Return row by index
    public function row($row) {
        return $this->m_cubies[$row];
    }
    
    // Return column by index
    public function column($column) {
        return $this->_transpose($this->m_cubies)[$column];
    }
    
    // Set slice row
    public function setRow(CubieVector $row, $idx) {
        $this->m_cubies[$idx] = $row;
        
        return $this;
    }
    
    // Set slice column
    public function setColumn(CubieVector $column, $idx) {
        $this->m_cubies = $this->_transpose($this->m_cubies);
        $this->m_cubies[$idx] = $column;
        $this->m_cubies = $this->_transpose($this->m_cubies);
    
        return $this;
    }
    
    // Rotate slice
    public function rotate($rot = Rotation::CLOCKWISE) {
        switch ($rot) {
            case Rotation::CLOCKWISE:
                $this->m_cubies = $this->_reverse($this->_transpose($this->m_cubies), Axis::HORIZONTAL);
                
                break;
            case Rotation::COUNTER_CLOCKWISE:
                $this->m_cubies = $this->_reverse($this->_transpose($this->m_cubies), Axis::VERTICAL);
                
                break;
            case Rotation::HALF:
                $this->m_cubies = $this->_reverse($this->_reverse($this->m_cubies, Axis::HORIZONTAL), Axis::VERTICAL);
                
                break;
        }
        
        return $this;
    }
    
    // Return formatted row by index
    public function formatRow($row) {
        $format = "";
        
        // Iterate cubies
        foreach ($this->m_cubies[$row] as $cubie)
            $format .= " $cubie";
        
        return $format;
    }
    
    // Return formatted slice
    public function format() {
        $format = "";
        
        // Iterate rows
        foreach ($this->m_cubies as $index=>$row)
            $format .= $this->formatRow($index).PHP_EOL;
        
        return $format;
    }
    
    // Return slice state
    public function state() {
        $format = "";
        
        // Iterate rows
        foreach ($this->m_cubies as $row)
            $format .= implode($row->getArrayCopy());
        
        return $format;
    }
    
    // Return transposed matrix
    public function _transpose(CubieMatrix $cubies) {
        $transposedCubies = clone $cubies;
        
        for ($i = 0; $i < count($cubies); ++$i)
            for ($j = 0; $j < count($cubies[$i]); ++$j)
                $transposedCubies[$j][$i] = $cubies[$i][$j];
        
        return $transposedCubies;
    }
    
    // Return reversed (axis-flipped) matrix
    public function _reverse(CubieMatrix $cubies, $axis) {
        $width = count($cubies);
        $reversedCubies = clone $cubies;
        
        switch ($axis) {
            case Axis::HORIZONTAL:
                for ($i = 0; $i < $width; ++$i)
                    for ($j = 0; $j < $width; ++$j)
                        $reversedCubies[$i][$width - $j - 1] = $cubies[$i][$j];
                
                break;
            case Axis::VERTICAL:
                for ($i = 0; $i < $width; ++$i)
                    for ($j = 0; $j < $width; ++$j)
                        $reversedCubies[$width - $i - 1][$j] = $cubies[$i][$j];
                
                break;
        }
        
        return $reversedCubies;
    }
}
?>